py/objstringio.h

```diff
@@ -33,6 +33,8 @@ typedef struct _mp_obj_stringio_t {
     vstr_t *vstr;
     // StringIO has single pointer used for both reading and writing
     mp_uint_t pos;
+    // Underlying object buffered by this StringIO
+    mp_obj_t ref_obj;
```

Comment:
> Is the sole purpose of this to prevent the underlying object from being GC'd?

Yes, it's done with that in mind, but also to cover what's described in https://github.com/micropython/micropython/issues/2622#issuecomment-275281602 - getvalue() would cast BytesIO's internal vstr to bytes, and store it there, subsequent calls will return it, and if code starts to write to it again, it will be COWed by the code added by this patch transparently. That of course means there's extra pointer to keep that string alive. But to get around that, there would be needed a dedicated method like .stealvalue() as discussed by the link above. Otherwise, we can just go "heuristic" method of clearing such cached string on .close/.truncate.